### ISSUE TITLE
Copy page tree with apphooks throws a server error

### ISSUE THREAD

**Issue by fsbraun** [django-cms/django-cms#5899]
### Summary
When trying to __copy a page tree__ that somewhere down the tree has an __apphook__ installed and then pasting the page tree somewhere else, Django-CMS throws a server error. 

It seems that copying apphooks is not implemented and if I try copying a single page w/ apphook I get the appropriate message ``Error: This page cannot be copied because an application is attached to it. See the Page's Advanced settings to manage apphooks.``

I consider this __important__, since the __process for editing an entire page tree__ is to first copy it, make the modifications (while the original tree is still online) and then publish the copy and unpublish the original tree.

### Expected behaviour
I would expect a meaningful error message instead of the 500 server error.

Even better, I would like Django-CMS __copy the full page tree__ w/ apphooks. If the latter is not possible for whatever technical reason, I would like to at least __get the full tree with apphooks removed__ plus a message that they could not be copied.

#### Actual behaviour
Error message ``Error: INTERNAL SERVER ERROR``

#### Environment

* Python version: 3.6
* Django version: 1.8.14
* django CMS version: 3.4.2

#### Thanks
Great work, everyone! Thanks very much!

**Comment by czpython** [comment]
Hello @fsbraun,
Thanks for reporting.

As you mentioned, there's two parts to this.
One is the db error that is raised when copying a page that contains another page with apphook and the other is the javascript error raised when copying a single page with an apphook.

I think in both cases, the only solution will be to remove the apphook from the new page.
/cc @divio/django-cms-core

**Comment by fsbraun** [comment]
Agree. I am not into the details of the DjangoCMS machine but the issue is that the namespace field of each apphook instance needs to be unique. So it is better to remove the apphook. 

This could be done by inserting

    if page.application_namespace:
        page.application_namespace =None
        page.application_urls = None
        page.navigation_extenders = None

before the ``page.save()`` in [line 491 of ``pagemodel.py``](https://github.com/divio/django-cms/blob/release/3.4.x/cms/models/pagemodel.py#L491).

Potentially, a message should be sent to the user notifying him/her of the removed appshooks.

**Comment by Srikar-V675** [comment]
Hi @fsbraun, 

I would like to work on this issue and I'm new to open source contributions, but I have worked with git. Can you clarify as to how to go ahead with this issue and what the potential solution could be?

Best regards,
Srikar V.

**Comment by fsbraun** [comment]
@Srikar-V675 Hey, that's good news. 

The issue has aged a bit, but I think it probably is still valid. So, since you've found an issue to work on, you should familiarize yourself with the contribution guidelines (https://docs.django-cms.org/en/latest/contributing/index.html). 

Once you have pinpointed the issue, follow the process for starting your contribution journey:
* [Forking the repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) 
* [Cloning the fork](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository)
* Making your changes locally using your preferred editor
* Updating and running tests locally
* [Committing changes](https://docs.github.com/en/desktop/making-changes-in-a-branch/committing-and-reviewing-changes-to-your-project-in-github-desktop)
* [Submitting a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request)

Since this is a regression, we will definitively need a regression test to make sure it has been fixed. 

Here https://github.com/django-cms/django-cms/issues/5899#issuecomment-292469015 I have posted a suggestion on how to fix the issue. **I would suggest you to do the fix on the django CMS 3.11 branch, since its code base is closer to the one I looked into in 2017:**  https://github.com/django-cms/django-cms/tree/release/3.11.x . Later, we can port forward the fix to the latest development version.

**Comment by Srikar-V675** [comment]
Hi @fsbraun, 

Since I'm new to open source contributions and the django-cms's code base. I would like to spend some time familiarizing with the code base and identifying the issue. I will make sure to give you some updates regularly.

Also I'm not quite aware of what you mean by a regression test.

**Comment by fsbraun** [comment]
Sure, @Srikar-V675 ! Can I invite you to our [contributor discord channel](https://discord-contributors-channel.django-cms.org/ 
). It's an excellent place to ask questions should any come up. 

A regression test is a test that fails when the bug (regression) is present and passes when it's fixed. This ensures that the bug is not reintroduced later. In this case it would be a test that copies an apphooked page and asserts that the apphook is not copied.

**Comment by Srikar-V675** [comment]
Thanks for the prompt response @fsbraun !

I'll take up your invite and join the discord channel and make sure to ask questions if any problems pop up. Thanks for the clarification regarding regression test.

**Comment by Srikar-V675** [comment]
Hi @fsbraun,

How do I replicate this issue? Are there any specific things I should be wary of?

**Comment by Vincent-Ngobeh** [comment]
Hi, I'd like to work on this issue.
I've reviewed the discussion and plan to implement the fix by clearing the apphook-related fields (application_urls, application_namespace, navigation_extenders) in the `Page.copy()` method when copying pages. This will allow page trees containing apphook pages to be copied without causing server errors.
I'll also add a regression test to ensure this doesn't break in the future.
I'll submit a PR against the `main` branch shortly.
